go/interpreter/lfvm/stack.go

```diff
@@ -11,85 +11,113 @@
-	stack_ptr int
+// stack is the 1024-element 256-bit word-wide stack used by the VM.
+// It is a fixed-size stack to prevent memory reallocation during execution.
+// Check boundaries are not checked. Users of the stack must prevent over- and
+// underflow situations.
+//
+// Each stack consumes 1024 * 32 bytes = 32KB of memory. Thus, creating and
+// destroying stacks could incur significant overhead. To mitigate this, a
+// stack pool is provided to reuse stack instances. To obtain an empty stack
+// from the pool, use NewStack(). To return a stack to the pool, use
+// ReturnStack(s).
+//
+// Example usage:
+//
+//	s := NewStack()
+//	defer ReturnStack(s)
+//	<use the stack in your local scope>
+//
+// The stack is not thread-safe. NewStack() and ReturnStack() are thread-safe.
+type stack struct {
+	data         [maxStackSize]uint256.Int
+	stackPointer int
 }
 
-func init() {
-	for i := 0; i < int(NUM_OPCODES); i++ {
-		staticStackBoundary[OpCode(i)] = getStaticStackInternal(OpCode(i))
-	}
+// push adds a copy of the given value to the top of the stack.
+func (s *stack) push(d *uint256.Int) {
+	s.data[s.stackPointer] = *d
+	s.stackPointer++
 }
 
-func (s *Stack) Data() []uint256.Int {
-	return s.data[:s.stack_ptr]
+// push adds a value with an undefined value to the top of the stack and returns
+// a pointer to this element. Use this function if the element on the top stack
+// should be modified directly using the returned pointer.
+func (s *stack) pushUndefined() *uint256.Int {
+	s.stackPointer++
+	return &s.data[s.stackPointer-1]
 }
 
-func (s *Stack) push(d *uint256.Int) {
-	s.data[s.stack_ptr] = *d
-	s.stack_ptr++
+// pop removes the top element from the stack and returns a pointer to it. The
+// obtained pointer is only valid until the next push operation. The pointer
+// can be used to obtain the pushed element without the need to copy it.
+func (s *stack) pop() *uint256.Int {
+	s.stackPointer--
+	return &s.data[s.stackPointer]
 }
 
-func (s *Stack) pushEmpty() *uint256.Int {
-	s.stack_ptr++
-	return &s.data[s.stack_ptr-1]
+// peek returns a pointer to the top element of the stack without removing it.
+// The returned pointer is only valid until the next operation on the stack.
+func (s *stack) peek() *uint256.Int {
+	return &s.data[s.len()-1]
 }
 
-func (s *Stack) pop() *uint256.Int {
-	s.stack_ptr--
-	return &s.data[s.stack_ptr]
+// peekN returns a pointer to the n-th element from the top of the stack without
+// removing it. The top element is at index 0 Thus, peekN(0) is equivalent to
+// peek().
+func (s *stack) peekN(n int) *uint256.Int {
+	return &s.data[s.len()-n-1]
 }
 
-func (s *Stack) len() int {
-	return s.stack_ptr
+// len returns the number of elements on the stack.
+func (s *stack) len() int {
+	return s.stackPointer
 }
 
-func (s *Stack) swap(n int) {
+// swap exchanges the top element with the n-th element from the top. The top
+// element is at index 1. Thus, swap(1) is a no-op.
+func (s *stack) swap(n int) {
```

Comment:
would it not be more intuitive that `swap(0)` is the no-op ? since the documentation says "top element" in `peekN` it says   "The top element is at index 0 Thus, peekN(0) is equivalent to peek().", maybe we should try to have some consistency between methods.

Reply:
Done. I also updated `dup` to start at index 0 to have all functions aligned.